front/pages/api/apps/[user]/[sId]/runs/index.js

```diff
@@ -49,124 +51,190 @@ export default async function handler(req, res) {
+        userId: user.id,
+      },
+    }),
+  ]);
+
+  const runRes = await fetch(
+    `${DUST_API}/projects/${app.dustAPIProjectId}/runs/stream`,
+    {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify({
+        run_type: "local",
+        specification_hash: req.body.specificationHash,
+        inputs: req.body.inputs,
+        config: { blocks: JSON.parse(req.body.config) },
+        credentials: credentialsFromProviders(providers),
+      }),
+    }
+  );
+
+  if (!runRes.ok) {
+    const error = await runRes.json();
+    res.status(400).json(error.error);
+    return;
+  }
 
-      const run = await runRes.json();
+  res.writeHead(200, {
+    "Content-Type": "text/event-stream",
+    "Cache-Control": "no-cache",
+    Connection: "keep-alive",
+  });
 
-      await app.update({
-        savedSpecification: req.body.specification,
-        savedConfig: req.body.config,
-        savedRun: run.response.run.run_id,
-      });
+  try {
+    for await (const chunk of runRes.body) {
+      res.write(chunk);
+      res.flush();
+    }
+  } catch (e) {
+    console.log("ERROR streaming from Dust API", e);
+  }
+  res.end();
+  return;
+}
 
-      res.status(200).json({ run: run.response.run });
-      break;
+async function designModeHandler(req, res, app, user) {
+  let [providers] = await Promise.all([
+    Provider.findAll({
+      where: {
+        userId: user.id,
+      },
+    }),
+  ]);
 
-    case "GET":
-      let limit = req.query.limit ? parseInt(req.query.limit) : 10;
-      let offset = req.query.offset ? parseInt(req.query.offset) : 0;
-      let runType = req.query.runType ? req.query.runType : "local";
-
-      const runsRes = await fetch(
-        `${DUST_API}/projects/${app.dustAPIProjectId}/runs?limit=${limit}&offset=${offset}&run_type=${runType}`,
-        {
-          method: "GET",
-        }
-      );
-
-      if (!runsRes.ok) {
-        const error = await runsRes.json();
-        res.status(400).json(error.error);
-        break;
-      }
+  if (
+    !req.body ||
+    !(typeof req.body.config == "string") ||
+    !(typeof req.body.specification === "string")
+  ) {
+    res.status(400).end();
+    return;
+  }
 
-      const runs = await runsRes.json();
+  const datasetsRes = await fetch(
+    `${DUST_API}/projects/${app.dustAPIProjectId}/datasets`,
+    {
+      method: "GET",
+    }
+  );
+  const datasets = await datasetsRes.json();
+  if (datasets.error) {
+    res.status(500).end();
+    return;
+  }
 
-      res
-        .status(200)
-        .json({ runs: runs.response.runs, total: runs.response.total });
-      break;
+  let latestDatasets = {};
+  for (const d in datasets.response.datasets) {
+    latestDatasets[d] = datasets.response.datasets[d][0].hash;
+  }
 
-    default:
-      res.status(405).end();
-      break;
+  const config = JSON.parse(req.body.config);
+  const { dataset: inputDataset } = Object.values(config).find(
```

Comment:
Errors if there is not `input` block in the app:
```
Executing (default): SELECT "id", "providerId", "config", "createdAt", "updatedAt", "userId" FROM "providers" AS "provider" WHERE "provider"."userId" = 1;
error - pages/api/apps/[user]/[sId]/runs/index.js (199:19) @ inputDataset
error - TypeError: Cannot destructure property 'dataset' of 'Object.values(...).find(...)' as it is undefined.
    at designModeHandler (webpack-internal:///(api)/./pages/api/apps/[user]/[sId]/runs/index.js:171:22)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Object.apiResolver (/home/spolu/code/dust/front/node_modules/next/dist/server/api-utils/node.js:372:9)
    at async DevServer.runApi (/home/spolu/code/dust/front/node_modules/next/dist/server/next-server.js:514:9)
    at async Object.fn (/home/spolu/code/dust/front/node_modules/next/dist/server/next-server.js:828:35)
    at async Router.execute (/home/spolu/code/dust/front/node_modules/next/dist/server/router.js:243:32)
    at async DevServer.runImpl (/home/spolu/code/dust/front/node_modules/next/dist/server/base-server.js:432:29)
    at async DevServer.run (/home/spolu/code/dust/front/node_modules/next/dist/server/dev/next-dev-server.js:831:20)
    at async DevServer.handleRequestImpl (/home/spolu/code/dust/front/node_modules/next/dist/server/base-server.js:375:20)
    at async /home/spolu/code/dust/front/node_modules/next/dist/server/base-server.js:157:99 {
  page: '/api/apps/[user]/[sId]/runs'
}
  197 | 
  198 |   const config = JSON.parse(req.body.config);
> 199 |   const { dataset: inputDataset } = Object.values(config).find(
      |                   ^
  200 |     (configValue) => configValue.type == "input"
  201 |   );
  202 | 
```

Reply:
Why did you change the logic here ooc?

Reply:
Original logic was a bit harder to read. But I forgot `|| {}`, my bad. Fixed